gui/packages/desktop/src/shared/ipc-event-channel.js

```diff
@@ -0,0 +1,155 @@
+// @flow
+
+import { ipcMain, ipcRenderer } from 'electron';
+import type { WebContents } from 'electron';
+
+import type { AppUpgradeInfo, CurrentAppVersionInfo } from '../main/index';
+import type { Location, RelayList, Settings, TunnelStateTransition } from '../main/daemon-rpc';
+
+export type AppStateSnapshot = {
+  isConnected: boolean,
+  tunnelState: TunnelStateTransition,
+  settings: Settings,
+  location: ?Location,
+  relays: RelayList,
+  currentVersion: CurrentAppVersionInfo,
+  upgradeVersion: AppUpgradeInfo,
+};
+
+interface Sender<T> {
+  notify(newState: T): void;
+}
+
+interface Receiver<T> {
+  listen<T>(fn: (T) => void): void;
+}
+
+/// Events names
+
+const DAEMON_CONNECTED = 'daemon-connected';
+const DAEMON_DISCONNECTED = 'daemon-disconnected';
+const TUNNEL_STATE_CHANGED = 'tunnel-state-changed';
+const SETTINGS_CHANGED = 'settings-changed';
+const LOCATION_CHANGED = 'location-changed';
+const RELAYS_CHANGED = 'relays-changed';
+const CURRENT_VERSION_CHANGED = 'current-version-changed';
+const UPGRADE_VERSION_CHANGED = 'upgrade-version-changed';
+
+/// Typed IPC event channel
+///
+/// Static methods are meant to be provide the way to send the events from a renderer process, while
+/// instance methods are meant to be used from a main process.
+///
+export default class IpcEventChannel {
+  _webContents: WebContents;
+
+  constructor(webContents: WebContents) {
+    this._webContents = webContents;
+  }
+
+  static state = {
+    /// Should be used from the main process to process state snapshot requests
+    serve(fn: () => AppStateSnapshot) {
+      ipcMain.on('get-state', (event) => {
+        event.returnValue = fn();
+      });
+    },
+
+    /// Synchronously sends the IPC request and returns the app state snapshot
+    get(): AppStateSnapshot {
+      return ipcRenderer.sendSync('get-state');
+    },
+  };
+
+  static daemonConnected: Receiver<void> = {
+    listen: listen(DAEMON_CONNECTED),
+  };
+
+  get daemonConnected(): Sender<void> {
+    return {
+      notify: sender(this._webContents, DAEMON_CONNECTED),
+    };
+  }
+
+  static daemonDisconnected: Receiver<?string> = {
+    listen: listen(DAEMON_DISCONNECTED),
+  };
+
+  get daemonDisconnected(): Sender<?string> {
+    return {
```

Comment:
FYI Since it's another `Object`, `this` is not preserved between calls, so I need to use "getter" in order to capture `this._webContents`.